aw_watcher_window/macos.swift

```diff
@@ -0,0 +1,285 @@
+        CFRunLoopMode.defaultMode)
+    }
+
+    let frontmost = NSWorkspace.shared.frontmostApplication!
+    let pid = frontmost.processIdentifier
+    let focusedApp = AXUIElementCreateApplication(pid)
+
+    AXObserverCreate(
+      pid,
+      {
+        (
+          _ axObserver: AXObserver,
+          axElement: AXUIElement,
+          notification: CFString,
+          userData: UnsafeMutableRawPointer?
+        ) -> Void in
+        guard let userData = userData else {
+          print("Missing userData")
+          return
+        }
+        let application = Unmanaged<MainThing>.fromOpaque(userData).takeUnretainedValue()
+        if notification == kAXFocusedWindowChangedNotification as CFString {
+          application.focusedWindowChanged(axObserver, window: axElement)
+        } else {
+          application.windowTitleChanged(
+            axObserver, axElement: axElement, notification: notification)
+        }
+      }, &observer)
+
+    let selfPtr = UnsafeMutableRawPointer(Unmanaged.passUnretained(self).toOpaque())
+    AXObserverAddNotification(
+      observer!, focusedApp, kAXFocusedWindowChangedNotification as CFString, selfPtr)
+
+    CFRunLoopAddSource(
+      RunLoop.current.getCFRunLoop(),
+      AXObserverGetRunLoopSource(observer!),
+      CFRunLoopMode.defaultMode)
+
+    var focusedWindow: AnyObject?
+    AXUIElementCopyAttributeValue(focusedApp, kAXFocusedWindowAttribute as CFString, &focusedWindow)
+
+    if focusedWindow != nil {
+      focusedWindowChanged(observer!, window: focusedWindow as! AXUIElement)
+    }
+  }
+}
+
+func checkAccess() -> Bool {
+  let checkOptPrompt = kAXTrustedCheckOptionPrompt.takeUnretainedValue() as NSString
+  let options = [checkOptPrompt: true]
+  let accessEnabled = AXIsProcessTrustedWithOptions(options as CFDictionary?)
+  return accessEnabled
+}
+
+func SystemIdleTime() -> Double? {
+  var iterator: io_iterator_t = 0
+  defer { IOObjectRelease(iterator) }
+  guard
+    IOServiceGetMatchingServices(kIOMainPortDefault, IOServiceMatching("IOHIDSystem"), &iterator)
+      == KERN_SUCCESS
+  else {
+    return nil
+  }
+
+  let entry: io_registry_entry_t = IOIteratorNext(iterator)
+  defer { IOObjectRelease(entry) }
+  guard entry != 0 else { return nil }
+
+  var unmanagedDict: Unmanaged<CFMutableDictionary>? = nil
+  defer { unmanagedDict?.release() }
+  guard
+    IORegistryEntryCreateCFProperties(entry, &unmanagedDict, kCFAllocatorDefault, 0) == KERN_SUCCESS
+  else { return nil }
+  guard let dict = unmanagedDict?.takeUnretainedValue() else { return nil }
+
+  let key: CFString = "HIDIdleTime" as CFString
+  let value = CFDictionaryGetValue(dict, Unmanaged.passUnretained(key).toOpaque())
+  let number: CFNumber = unsafeBitCast(value, to: CFNumber.self)
+  var nanoseconds: Int64 = 0
+  guard CFNumberGetValue(number, CFNumberType.sInt64Type, &nanoseconds) else { return nil }
+  let interval = Double(nanoseconds) / Double(NSEC_PER_SEC)
+
+  return interval
+}
```

Comment:
We already have this in aw-watcher-afk, so not needed here: https://github.com/ActivityWatch/aw-watcher-afk/blob/master/aw_watcher_afk/macos.py

However, I'd be interested in any differences in behavior the implementations might have (I can't really tell).

Reply:
This has two differences: 
1) it uses the iokit framework which is more modern (and more complicated) to detect time since last idle
2) It is optimized for efficiently (low cpu) getting idle time and unidle time down to the ms. This means it doesn't poll every second, instead it sleeps for the exact amount of time before idle (currently set to 15 seconds, but could be set to any number and it will only check when it needs to), and then once idle registers a callback that will detect wakeups immediately